### ISSUE TITLE
missed optimization under the condition that modulo (a % b) implies

### ISSUE THREAD

**Issue by k-arrows** [llvm/llvm-project#73422]
Test case:
https://godbolt.org/z/Kh7zcrvoz
```cpp
int foo(int a) {
  if ((a % 5) == 3)
    return (a < 0);
  else
    return 0;
}
```
Can CVP handle this?
See Also: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=91029

**Comment by XChy** [comment]
Similar to #71533, this optimization requires making use of KnownBits information inside select arms in InstCombine. That's probably technically hard to solve now.

**Comment by k-arrows** [comment]
> Similar to #71533, this optimization requires making use of KnownBits information inside select arms in InstCombine. That's probably technically hard to solve now.

Thank you for your comment. I used modulo as an example in this issue, but I'm sure there are other examples where making use of KnownBits information might be useful.

Example of `x & MASK`:
https://godbolt.org/z/8fcjhjqcd
(See Also: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=107043)
```cpp
int test(int n)
{
  if ((n & 0x8000) == 0)
    return 1;
  return (n >> 15) & 0x1;
}
```

**Comment by XChy** [comment]
Yes, LLVM tends to fold simple if-else to a select. But InstCombine is insensitive to its context on select arms, requiring refactoring related logic. I may give it a try for simple `icmp` later, not sure whether a good solution...

**Comment by k-arrows** [comment]
> Example of `x & MASK`: https://godbolt.org/z/8fcjhjqcd (See Also: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=107043)

This example has been fixed on trunk.
